Character set and collation updates (#19976)
pingcap/docs: character-set-and-collation.md

1 row in set (0.00 sec)
```

> The following demonstates differences in how various Unicode collations compare the `ß` that is used in the German language to `ss`. Here you can see that only the more strict Unicode collations consider this to be the same, resulting in `1` (which means `TRUE`).

Reviewer: ```suggestion
The following demonstrates how different Unicode collations compare the German `ß` with `ss`. You can see that only the more strict Unicode collations treat them as equivalent, returning 1 (which means `TRUE`).
```

```sql
SELECT
  'ss' COLLATE utf8mb4_general_ci = 'ß',
  'ss' COLLATE utf8mb4_unicode_ci = 'ß',
  'ss' COLLATE utf8mb4_0900_ai_ci = 'ß',
  'ss' COLLATE utf8mb4_0900_bin = 'ß'
\G
```

```
*************************** 1. row ***************************
'ss' COLLATE utf8mb4_general_ci = 'ß': 0
'ss' COLLATE utf8mb4_unicode_ci = 'ß': 1
'ss' COLLATE utf8mb4_0900_ai_ci = 'ß': 1
  'ss' COLLATE utf8mb4_0900_bin = 'ß': 0
1 row in set (0.01 sec)
```

> ### Character set and collation naming

Reviewer: ```suggestion
### Suffixes of character sets and collations
```

Author: Maybe we should keep the original title and add this:

```suggestion
### Character set and collation naming

One character set can have many collations. The collations are named `<character_set>_<collation_properties>`. For example the `utf8mb4` character set has a collation named `utf8mb4_bin`, which is a binary collation for the `utf8mb4` character set.

```

1 row in set (0.00 sec)
```

The following demonstrates how different Unicode collations compare the German `ß` with `ss`. You can see that only the more strict Unicode collations treat them as equivalent, returning 1 (which means `TRUE`).

```sql
SELECT
  'ss' COLLATE utf8mb4_general_ci = 'ß',
  'ss' COLLATE utf8mb4_unicode_ci = 'ß',
  'ss' COLLATE utf8mb4_0900_ai_ci = 'ß',
  'ss' COLLATE utf8mb4_0900_bin = 'ß'
\G
```

```
*************************** 1. row ***************************
'ss' COLLATE utf8mb4_general_ci = 'ß': 0
'ss' COLLATE utf8mb4_unicode_ci = 'ß': 1
'ss' COLLATE utf8mb4_0900_ai_ci = 'ß': 1
  'ss' COLLATE utf8mb4_0900_bin = 'ß': 0
1 row in set (0.01 sec)
```

### Character set and collation naming

A character set can have multiple collations, named in the format `<character_set>_<collation_properties>`. For example, the `utf8mb4` character set has a collation called `utf8mb4_bin`, which is a binary collation for `utf8mb4`. Multiple collation properties can be included in the name, separated by underscores (`_`). 

> The following table shows the suffixes and meanings of character sets and collations.

Reviewer 2: ```suggestion
The following table shows the common collation properties and meanings.
```

> | Suffix | Meaning |

Reviewer 2: ```suggestion
| Collation properties | Meaning |
```

|---|---|

> | `_bin` | binary |
> | `_ci` | case insensitive |
> | `_ai_ci` | accent insensitive, case insensitive |
> | `_0900_bin` | Unicode UCA 9.0.0, binary |
> | `_unicode_ci` | (older) Unicode UCA collation, case insensitive |

Reviewer 2: ```suggestion
| `_bin` | Binary |
| `_ci` | Case insensitive |
| `_ai_ci` | Accent insensitive, case insensitive |
| `_0900_bin` | Unicode UCA 9.0.0, binary |
| `_unicode_ci` | (Older) Unicode UCA collation, case insensitive |
```

## `utf8` and `utf8mb4` in TiDB

> In MySQL, the character set `utf8` is limited to a maximum of three bytes. This is sufficient to store characters in the Basic Multilingual Plane (BMP), but not enough to store characters such as emojis.  For new installations, it is recommended to use `utf8mb4` and migrate away from `utf8`.

Reviewer: ```suggestion
In MySQL, the character set `utf8` is limited to a maximum of three bytes. This is sufficient to store characters in the Basic Multilingual Plane (BMP), but not enough to store characters such as emojis. For new installations, it is recommended to use `utf8mb4` and migrate away from `utf8`.
```

> In MySQL, the character set `utf8` is limited to a maximum of three bytes. This is sufficient to store characters in the Basic Multilingual Plane (BMP), but not enough to store characters such as emojis. It is recommended to use the character set `utf8mb4` for new installations and migrate away from `utf8`.

Reviewer: ```suggestion
In MySQL, the character set `utf8` is limited to a maximum of three bytes. This is sufficient to store characters in the Basic Multilingual Plane (BMP), but not enough to store characters such as emojis.  For new installations, it is recommended to use `utf8mb4` and migrate away from `utf8`.
```

> In both MySQL and TiDB `utf8` and `utf8mb3` are aliases for the same character set.

Reviewer: ```suggestion
In both MySQL and TiDB, `utf8` and `utf8mb3` are aliases for the same character set.
```

## `utf8` and `utf8mb4` in TiDB

In MySQL, the character set `utf8` is limited to a maximum of three bytes. This is sufficient to store characters in the Basic Multilingual Plane (BMP), but not enough to store characters such as emojis. For new installations, it is recommended to use `utf8mb4` and migrate away from `utf8`.

In both MySQL and TiDB, `utf8` and `utf8mb3` are aliases for the same character set.

> By default, TiDB also limits the character set `utf8` to a maximum of three bytes to ensure that data created in TiDB can still safely be restored in MySQL. You can disable it by changing the value of the system variable [`tidb_check_mb4_value_in_utf8`](/system-variables.md#tidb_check_mb4_value_in_utf8) to `OFF`. However it is recommended to use `utf8mb4` instead.

Reviewer 2: ```suggestion
By default, TiDB also limits the character set `utf8` to a maximum of three bytes to ensure that data created in TiDB can still safely be restored in MySQL. You can disable it by changing the value of the system variable [`tidb_check_mb4_value_in_utf8`](/system-variables.md#tidb_check_mb4_value_in_utf8) to `OFF`. However, it is recommended to use `utf8mb4` instead for full Unicode support and better compatibility.
```
